index.js

```diff
@@ -107,55 +113,93 @@ exports = module.exports = function serveIndex(root, options){
-    if (~path.indexOf('\0')) return next(createError(400));
+    // content-negotiation
+    var accept = accepts(req);
+    var type = accept.type(mediaTypes);
 
-    // malicious path
-    if ((path + sep).substr(0, root.length) !== root) {
-      debug('malicious path "%s"', path);
-      return next(createError(403));
-    }
+    // not acceptable
+    if (!type) return next(createError(406));
 
-    // determine ".." display
-    var showUp = normalize(resolve(path) + sep) !== root;
+    var batch = new Batch();
 
-    // check if we have a directory
-    debug('stat "%s"', path);
-    fs.stat(path, function(err, stat){
-      if (err && err.code === 'ENOENT') {
-        return next();
-      }
+    batch.concurrency(10);
 
-      if (err) {
-        err.status = err.code === 'ENAMETOOLONG'
-          ? 414
-          : 500;
-        return next(err);
-      }
+    // read all root dirs and files with batch
+    roots.forEach(function(root){
+      batch.push(function(done){
 
-      if (!stat.isDirectory()) return next();
+        // join / normalize from root dir
+        var path = normalize(join(root, dir));
 
-      // fetch files
-      debug('readdir "%s"', path);
-      fs.readdir(path, function(err, files){
-        if (err) return next(err);
-        if (!hidden) files = removeHidden(files);
-        if (filter) files = files.filter(function(filename, index, list) {
-          return filter(filename, index, list, path);
+        // null byte(s), bad request
+        if (~path.indexOf('\0')) return done(createError(400));
+
+        // malicious path
+        if ((path + sep).substr(0, root.length) !== root) {
+          debug('malicious path "%s"', path);
+          return done(createError(403));
+        }
+
+        // check if we have a directory
+        debug('stat "%s"', path);
+        fs.stat(path, function(err, stat){
+          if (err && err.code === 'ENOENT') {
+            return done(null, null);
+          }
+
+          if (err) {
+            err.status = err.code === 'ENAMETOOLONG'
+              ? 414
+              : 500;
+            return done(err);
+          }
+
+          if (!stat.isDirectory()) return done(null, null);
+
+          // fetch files
+          debug('readdir "%s"', path);
+          fs.readdir(path, function(err, files){
+            if (err) return done(err);
+
+            // return dir path and files
+            done(null, { path: path, files: files });
+          });
+        });
+      });
+    });
+
+    // process all files
+    batch.end(function(err, dirs){
+      if (err) return next(err);
+
+      var files = [],   // all files
+          fileMap = {}; // remember already added files
+      dirs.forEach(function(dir){
+        if (!dir) return;
+        dir.files.forEach(function(file){
+          if (!fileMap[file]) {
+            fileMap[file] = true; // set flag to not add the file again
+            files.push({ name: file, path: dir.path });
```

Comment:
How does this handle URL case-sensitivity vs file system case-sensitivity? As in, what happens when you are on a case-insensitive file system one the two overlayed dirs have the same file: one is `foo.txt` and the other is `Foo.txt`. It seems like this would cause both to show in the UI, but of course both `foo.txt` and `Foo.txt` when requested in the URL is going to end up being the same one instead of different ones.